### PR TITLE
[FFI] Perf improvement via NativeMethodHandle/linkToNative

### PR DESCRIPTION
**Pull request by ChengJin01**
The changes aim to generate the NativeMethodHandle instance to ensure the invocation path goes via linkToNative.

Signed-off-by: ChengJin01 <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
The changes are the same as https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/736.

File: src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java [hunk 1]
```diff
@@ -25,15 +25,14 @@
 
 /*
  * ===========================================================================
- * (c) Copyright IBM Corp. 2022, 2022 All Rights Reserved
+ * (c) Copyright IBM Corp. 2022, 2023 All Rights Reserved
```
Comment (keithc-ca) [inline src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:28]: Copyright dates should be 2024.

Comment (ChengJin01) [reply src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java]: Fixed everywhere.

File: src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java [hunk 2]
```diff
@@ -25,15 +25,14 @@
...
- * (c) Copyright IBM Corp. 2022, 2022 All Rights Reserved
+ * (c) Copyright IBM Corp. 2022, 2024 All Rights Reserved
  * ===========================================================================
  */
 
 package java.lang.invoke;
 
 import jdk.internal.access.JavaLangInvokeAccess;
 import jdk.internal.access.SharedSecrets;
-import jdk.internal.foreign.abi.NativeEntryPoint;
```
Comment (keithc-ca) [inline src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:36]: I think you should leave this alone; it isn't used, but there's no harm keeping it.

Comment (ChengJin01) [reply src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java]: Fixed.

File: src/java.base/share/classes/java/lang/invoke/NativeMethodHandle.java [hunk 3]
```diff
@@ -74,11 +89,13 @@ private static boolean hasIllegalType(MethodType type) {
...
-              || pType == double.class
-              || pType == void.class
-              || pType == Object.class);
+                || pType == int.class
+                || pType == float.class
+                || pType == double.class
+                || pType == void.class
+                || pType == Object.class
+                || pType == MemorySegment.class // Skip checking the argument preset in OpenJ9 for the moment.
+                || pType == SegmentAllocator.class); // Skip checking the argument preset OpenJ9 for the moment.
```
Comment (keithc-ca) [inline src/java.base/share/classes/java/lang/invoke/NativeMethodHandle.java:98]: Please don't change the indentation.

Comment (ChengJin01) [reply src/java.base/share/classes/java/lang/invoke/NativeMethodHandle.java]: Fixed.

File: src/java.base/share/classes/java/lang/invoke/NativeMethodHandle.java [hunk 4]
```diff
@@ -65,6 +80,10 @@ private static boolean hasIllegalType(MethodType type) {
             return true;
 
         for (Class<?> pType : type.ptypes()) {
+          /* Specifically allow these parameter types for OpenJ9 for the moment. */
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply src/java.base/share/classes/java/lang/invoke/NativeMethodHandle.java]: Fixed.

**Comment by keithc-ca** [comment]
Jenkins test sanity alinux jdk22 depends https://github.com/eclipse-openj9/openj9/pull/18799

**Comment by keithc-ca** [comment]
Jenkins test sanity alinux jdk22 depends https://github.com/eclipse-openj9/openj9/pull/18799

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**